services/src/projects/projectdb.rs

```diff
@@ -0,0 +1,383 @@
+use crate::projects::project::{OrderBy, ProjectFilter, ProjectListing, ProjectOwner, ProjectId, Project, CreateProject, UpdateProject, UserProjectPermission, ProjectPermission, ProjectListOptions};
+use crate::users::user::{UserIdentification, Validated};
+use std::collections::HashMap;
+use crate::error::Result;
+use crate::error;
+use snafu::ensure;
+
+pub trait ProjectDB: Send + Sync {
```

Comment:
The DBs could use a startup and shutdown method.
This way, you could even persist and load the HashMap DBs.
Furthermore, a future PostgreSQL DB could establish the connection.

You of course have a constructor for all these types, but I'm not sure if this suffices and if required methods help.
Then, there should be hooks for starting and stopping... this is a quite wild guess, what do you think?

Reply:
I think the constructor suffices for now, let's see what we need when we come to implementing a persistent storage